src/common/database.c

```diff
@@ -2618,49 +2663,147 @@ dt_database_t *dt_database_init(const char *alternative, const gboolean load_dat
       db = NULL;
 
-      if(shall_we_delete_the_db)
+      if(resp != GTK_RESPONSE_ACCEPT && resp != GTK_RESPONSE_REJECT)
       {
-        fprintf(stderr, "[init] deleting `%s' on user request", dbfilename_data);
+        fprintf(stderr, "[init] database `%s' is corrupt and can't be opened! either replace it from a backup or "
+        "delete the file so that darktable can create a new one the next time. aborting\n", dbfilename_data);
+        g_free(data_snap);
+        goto error;
+      }
 
-        if(g_unlink(dbfilename_data) == 0)
-          fprintf(stderr, " ... ok\n");
-        else
-          fprintf(stderr, " ... failed\n");
+      //here were sure that response is either accept (restore from snap) or reject (just delete the damaged db)
 
-        goto start;
-      }
+      fprintf(stderr, "[init] deleting `%s' on user request", dbfilename_data);
+
+      if(g_unlink(dbfilename_data) == 0)
+        fprintf(stderr, " ... ok\n");
       else
+        fprintf(stderr, " ... failed\n");
+
+      if(resp == GTK_RESPONSE_ACCEPT && data_snap)
       {
-        fprintf(stderr, "[init] database `%s' is corrupt and can't be opened! either replace it from a backup or "
-        "delete the file so that darktable can create a new one the next time. aborting\n", dbfilename_data);
-        goto error;
+        fprintf(stderr, "[init] restoring `%s' from `%s'...", dbfilename_data, data_snap);
+        GError *gerror = NULL;
+        if(!g_file_test(dbfilename_data, G_FILE_TEST_EXISTS))
+        {
+          GFile *src = g_file_new_for_path(data_snap);
+          GFile *dest = g_file_new_for_path(dbfilename_data);
+          gboolean copyStatus = TRUE;
+          if(g_file_test(data_snap, G_FILE_TEST_EXISTS))
+          {
+            copyStatus = g_file_copy(src, dest, G_FILE_COPY_NONE, NULL, NULL, NULL, &gerror);
+            if(copyStatus) copyStatus = g_chmod(dbfilename_data, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH) == 0;
+          }
+          else
+          {
+            // there is nothing to restore, create an empty file
+            int fd = g_open(dbfilename_data, O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+            if(fd < 0 || !g_close(fd, &gerror)) copyStatus = FALSE;
+          }
+          if(copyStatus)
+            fprintf(stderr, " success!\n");
+          else
+            fprintf(stderr, " failed!\n");
+
+          g_object_unref(src);
+          g_object_unref(dest);
+        }
       }
+      g_free(data_snap);
+      g_free(dbname);
+      goto start;
     }
   }
 
+  gchar* libdb_status = _get_pragma_string_val(db->handle, "main.quick_check");
   // next we are looking at the library database
   // does the db contain the new 'db_info' table?
   rc = sqlite3_prepare_v2(db->handle, "select value from main.db_info where key = 'version'", -1, &stmt, NULL);
-  if(rc == SQLITE_OK && sqlite3_step(stmt) == SQLITE_ROW)
+  if(!g_strcmp0(libdb_status, "ok") && rc == SQLITE_OK && sqlite3_step(stmt) == SQLITE_ROW)
```

Comment:
More a question, I suppose "ok" is what is returned in all languages ? I'm not even sure SQlite3 has translation...

Reply:
Yes. In all languages returned data is "ok" :) (I had to browse sqlite sources to check, but now I'm quite sure)

Reply:
Great then ! I'll test this ASAP.